xep-0388.xml

```diff
@@ -195,71 +300,187 @@
+  <example caption="SASL mechanism upgrade required"><![CDATA[
+<!-- The Server requests the client to perform the upgrade task for BLOOP2 -->
+<continue xmlns='urn:xmpp:sasl:2'>
+  <additional-data>
+    SSdtIGJvcmVkIG5vdy4=
+  </additional-data>
+  <tasks>
+    <task>UPGR-BLOOP2</task>
+  </tasks>
+  <text>This account requires an upgrade to BLOOP2 as requested by the client</text>
+</continue>
+
+<!-- The Client initiates the task -->
+<next xmlns='urn:xmpp:sasl:2' task='UPGR-BLOOP2'/>
+
+<!-- The Server provides the needed data -->
+<data xmlns='urn:xmpp:bloop2:example'>
+  <data>U28sIG5leHQgRk9TREVNIC0gMjAxOCwgdGhhdCBpcy4uLg==</data>
+  <description>BLOOP2 is cool!</description>
+</data>
+
+<!-- The Client now responds -->
+<upgrade xmlns='urn:xmpp:bloop2:example'>
+  <response>Li4uSSdsbCBidXkgYSBiZWVyIGZvciB0aGUgZmlyc3QgcGVyc29uIHdoby4uLg==</response>
+  <finalize>42</finalize>
+</upgrade>
+
+<!-- The upgrade was performed and the Server finishes authentication -->
+<success xmlns='urn:xmpp:sasl:2'>
+  <authorization-identifier>juliet@montague.example</authorization-identifier>
+</success>
+]]></example>
+
+  <section2 topic="SCRAM upgrade" anchor="scram-upgrade">
+    <p>For upgrades of SCRAM mechanisms as defined in &rfc5802;, the server has to provide the needed data for the client to calculate the SaltedPassword as defined in this RFC (or some RFC updating it), namely the iteration count and salt. To do so the server sends a &lt;parameters/> element namespaced to "urn:xmpp:sasl:2" containing the salt and iteration count in the elements &lt;salt/> and &lt;iteration/> as defined in that RFC, omitting the "s=" and "i=" prefix.</p>
+    <p>The client then calculates the SaltedPassword and sends back its base64 encoded value inside a &lt;hash/> element namespaced to "urn:xmpp:sasl:2".</p>
+    <p>The name of the upgrade task MUST NOT conain the "-PLUS" suffix, because channel-binding is not relevant for upgrade tasks.</p>
+    <example caption="SCRAM hash upgrade task for SCRAM-SHA-256 after successful SCRAM-SHA-1-PLUS authentication"><![CDATA[
+<!-- The Server provides the SCRAM upgrade task requested by the client in its <authenticate/> element -->
+<continue xmlns='urn:xmpp:sasl:2'>
+  <!-- Base64 of: 'v=msVHs/BzIOHDqXeVH7EmmDu9id8=' -->
+  <additional-data>
+    dj1tc1ZIcy9CeklPSERxWGVWSDdFbW1EdTlpZDg9
+  </additional-data>
+  <tasks>
+    <task>UPGR-SCRAM-SHA-256</task>
+  </tasks>
+</continue>
+
+<!-- The Client initiates the task -->
+<next xmlns='urn:xmpp:sasl:2' task='UPGR-SCRAM-SHA-256'/>
+
+<!--
+  The server sends the required salt and iteration count.
+-->
+<parameters xmlns='urn:xmpp:sasl:2'>
+  <salt>A_SXCRXQ6sek8bf_Z</salt>
+  <iterations>4096</iterations>
+</parameters>
+
+<!-- The client responds with the base64 encoded SaltedPassword. -->
+<hash xmlns='urn:xmpp:sasl:2'>
+  BzOnw3Pc5H4bOLlPZ/8JAy6wnTpH05aH21KW2+Xfpaw=
+</hash>
```

Comment:
See above - all worthy things, but baking a particular SASL mechanism into the SASL profile seems entirely counter to the principles of SASL.

Reply:
Well, as I said above I thought defining the SCRAM upgrade task was small addition and didn't want to introduce yet another new XEP, but of course we could move that to a new XEP.

Reply:
I'm moving the SCRAM part of upgrade tasks into a new ProtoXEP.